File: 222. Count Complete Tree Nodes/countNodesHeuristic.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {number}
 */
const countNodes = (root) => {
  if (!root) {
    return 0;
  }

  let left = root;
  let right = root;
  let leftDepth = 0;
  let rightDepth = 0;

  while (left) {
    leftDepth++;
    left = left.left
  }

  while (left) {
    rightDepth++;
    right = right.right
  }

  if (leftDepth === rightDepth) {
    return Math.pow(2, leftDepth) - 1;
  }

  return 1 + countNodes(root.left) + countNodes(root.right);
};

const root = new TreeNode(1);
root.left = new TreeNode(2);
root.right = new TreeNode(3);

root.left.left = new TreeNode(4);
root.left.right = new TreeNode(5);

root.right.left = new TreeNode(6);

console.log('countNodes: ', countNodes(root));

File: 222. Count Complete Tree Nodes/countNodesHeuristic.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def countNodes(self, root: Optional[TreeNode]) -> int:
    if not root:
      return 0
    
    left = root
    right = root
    leftDepth = 0
    rightDepth = 0
    
    while left:
      leftDepth += 1
      left = left.left
      
    while right:
      rightDepth += 1
      right = right.right
      
    if leftDepth == rightDepth:
      return pow(2, leftDepth) - 1
    
    return 1 + self.countNodes(root.left) + self.countNodes(root.right)


root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

root.left.left = TreeNode(4)
root.left.right = TreeNode(5)

root.right.left = TreeNode(6)

solution = Solution()
print('countNodes: ', solution.countNodes(root))
